src/main/java/com/cloudbees/hudson/plugins/folder/AbstractFolder.java

```diff
@@ -337,216 +334,73 @@ public static <K, V extends TopLevelItem> Map<K, V> loadChildren(AbstractFolder<
         final ChildNameGenerator<AbstractFolder<V>,V> childNameGenerator = parent.childNameGenerator();
         Map<String,V> byDirName = new HashMap<>();
         if (parent.items != null) {
-            if (childNameGenerator == null) {
-                for (V item : parent.items.values()) {
-                    byDirName.put(item.getName(), item);
-                }
-            } else {
-                for (V item : parent.items.values()) {
-                    String itemName = childNameGenerator.dirNameFromItem(parent, item);
-                    if (itemName == null) {
-                        itemName = childNameGenerator.dirNameFromLegacy(parent, item.getName());
-                    }
-                    byDirName.put(itemName, item);
-                }
+            for (V item : parent.items.values()) {
+                byDirName.put(childNameGenerator.dirName(parent, item), item);
             }
         }
         for (File subdir : subdirs) {
+            File effectiveSubdir = subdir;
+            String childName = childNameGenerator.readItemName(subdir);
+            // Try to retain the identity of an existing child object if we can.
+            V itemFromDir;
+            V item;
+            boolean itemNeedsSave = false;
+            String name;
+            item = itemFromDir = byDirName.get(childName);
```

Comment:
@vlatombe Note that the keys of `byDirName` have gone through `NameMangler` if they are an `OrganizationFolder` or `MultiBranchProjectDescriptor` with special characters. Before this PR, `childName` was reset to `subdir.getName()` after reading the child name file, so the lookup was correct. After this PR, the lookup is done with the unmangled name rather than the `NameMangler`-mangled name. Therefore, the attempt to "try to retain the identity of an existing child object if we can" fails. This is a regression.

Reply:
#486